File: src/services/LLMService.js
class LLMService {
  constructor() {
    this.apiKey = process.env.REACT_APP_GMI_API_KEY;
    this.apiUrl = process.env.REACT_APP_GMI_API_URL || 'https://api.gmi.cloud/v1/chat/completions';
  }

  async generateResponse(prompt, context = [], intent = 'general') {
    try {
      const systemPrompts = {
        analysis: 'You are DeepShield AI, a forensic expert in deepfake detection. Provide detailed, professional analysis of detection results in clear, educational language.',
        educational: 'You are DeepShield AI, an expert educator in AI security and deepfake detection. Explain complex concepts clearly and adapt to the user\'s technical level.',
        conversational: 'You are DeepShield AI, a helpful AI security expert. Provide professional yet approachable responses about deepfakes, AI security, and digital forensics.',
        threat_analysis: 'You are DeepShield AI, a cybersecurity expert specializing in deepfake threats. Provide comprehensive threat landscape analysis.'
      };

      const systemMessage = systemPrompts[intent] || systemPrompts.conversational;

      const response = await fetch(this.apiUrl, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          model: 'gpt-3.5-turbo',
          messages: [
            {
              role: 'system',
              content: systemMessage
            },
            ...context.slice(-6),
            {
              role: 'user',
              content: prompt
            }
          ],
          max_tokens: 600,
          temperature: 0.7
        })
      });

      if (!response.ok) {
        throw new Error(`API Error: ${response.status}`);
      }

      const data = await response.json();
      return data.choices[0].message.content;

    } catch (error) {
      console.error('LLM API Error:', error);
      return this.getFallbackResponse(prompt, intent);
    }
  }

  getFallbackResponse(prompt, intent) {
    const fallbacks = {
      analysis: "Based on comprehensive multi-method analysis using facial geometry, edge artifacts, texture patterns, and frequency domain signatures, this content shows the indicated confidence level for authenticity assessment.",
      educational: "I use advanced computer vision and machine learning techniques to detect deepfakes. My multi-method approach analyzes facial landmarks, edge patterns, texture consistency, and frequency domain artifacts.",
      threat_analysis: "Deepfakes pose significant threats to democratic processes, financial security, and personal safety. Early detection and verification systems are critical for maintaining digital trust.",
      conversational: "I'm DeepShield AI, your intelligent deepfake detection specialist. I can analyze media files and educate you about AI security threats. How can I help you today?"
    };

    return fallbacks[intent] || fallbacks.conversational;
  }

  async generateAnalysisExplanation(results, fileName, technicalLevel = 'intermediate') {
    const { confidence, isDeepfake, artifacts } = results;
    
    const prompt = `
Analysis Results for "${fileName}":
- Classification: ${isDeepfake ? 'DEEPFAKE DETECTED' : 'AUTHENTIC CONTENT'}
- Confidence Score: ${confidence.toFixed(1)}%
- Detection Methods: ${artifacts.map(a => a.type).join(', ')}
- Key Findings: ${artifacts.map(a => `${a.type}: ${a.score.toFixed(1)}%`).join('; ')}

User Technical Level: ${technicalLevel}

Generate a professional forensic analysis report explaining these findings. Include:
1. Executive summary of the results
2. Explanation of detection methods used
3. Key evidence found
4. Confidence assessment
5. Recommendations

Adapt the technical depth to the user's level: ${technicalLevel}.
`;

    return await this.generateResponse(prompt, [], 'analysis');
  }
}

export default new LLMService();

File: backend/llm_service.py
import os
import requests

API_KEY = os.getenv("GMI_API_KEY")
API_URL = os.getenv("GMI_API_URL", "https://api.gmi.cloud/v1/chat/completions")

SYSTEM_PROMPTS = {
    "analysis": "You are DeepShield AI, a forensic expert in deepfake detection. Provide detailed, professional analysis of detection results in clear, educational language.",
    "educational": "You are DeepShield AI, an expert educator in AI security and deepfake detection. Explain complex concepts clearly and adapt to the user's technical level.",
    "conversational": "You are DeepShield AI, a helpful AI security expert. Provide professional yet approachable responses about deepfakes, AI security, and digital forensics.",
    "threat_analysis": "You are DeepShield AI, a cybersecurity expert specializing in deepfake threats. Provide comprehensive threat landscape analysis."
}

FALLBACKS = {
    "analysis": "Based on multi-method forensic analysis, this content shows the indicated confidence level for authenticity assessment.",
    "educational": "Advanced computer vision and ML techniques were used, including facial landmarks, texture patterns, and frequency-based filters.",
    "threat_analysis": "Deepfakes pose threats to democracy, finance, and individual safety. Early detection is vital to protect digital ecosystems.",
    "conversational": "I'm DeepShield AI, your assistant in spotting and explaining deepfakes. Ask me anything about media security!"
}


def generate_response(prompt, intent="conversational", context=[]):
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json"
    }

    messages = [
        {"role": "system", "content": SYSTEM_PROMPTS.get(intent, SYSTEM_PROMPTS["conversational"])}
    ] + context[-6:] + [{"role": "user", "content": prompt}]

    body = {
        "model": "gpt-3.5-turbo",
        "messages": messages,
        "max_tokens": 600,
        "temperature": 0.7
    }

    try:
        response = requests.post(API_URL, headers=headers, json=body)
        response.raise_for_status()
        data = response.json()
        return data.get("choices", [{}])[0].get("message", {}).get("content", FALLBACKS[intent])
    except Exception as e:
        print("LLM error:", e)
        return FALLBACKS[intent]


def generate_analysis_explanation(result, filename, technical_level="intermediate"):
    confidence = f"{result.get('confidence', 0) * 100:.1f}"
    is_deepfake = "DEEPFAKE DETECTED" if result.get("label", "").lower() == "fake" else "AUTHENTIC CONTENT"
    artifacts = result.get("artifacts", [])

    artifact_summary = ", ".join(a.get("type", "unknown") for a in artifacts) or "None detected"
    artifact_details = "; ".join(f"{a.get('type')}: {a.get('score', 0)*100:.1f}%" for a in artifacts) or "No artifacts to report."

    prompt = (
        f"Analysis Results for \"{filename}\":\n"
        f"- Classification: {is_deepfake}\n"
        f"- Confidence Score: {confidence}%\n"
        f"- Detection Methods: {artifact_summary}\n"
        f"- Key Findings: {artifact_details}\n\n"
        f"User Technical Level: {technical_level}\n\n"
        f"Generate a professional forensic analysis report explaining these findings. Include:\n"
        f"1. Executive summary\n"
        f"2. Explanation of detection methods used\n"
        f"3. Key evidence found\n"
        f"4. Confidence assessment\n"
        f"5. Recommendations\n\n"
        f"Adapt the technical depth to the user's level: {technical_level}."
    )

    return generate_response(prompt, "analysis")
